File: qbopt/optimize/wholestores.py
"""Store a whole value instead of separately storing its extracted words."""

from dataclasses import replace

from qbopt.model import ir, mir


def _word(op):
    return (op.kind is mir.Kind.STORE and not op.barrier and not op.defines and not op.loads
            and len(op.args) == len(op.stores) == 1
            and isinstance(op.args[0], mir.Held) and op.args[0].width == 2
            and op.stores[0].width == 2 and op.stores[0].addr is not None)


def joined(body: mir.MirBody) -> mir.MirBody:
    definitions = {value: op for block in body.blocks for op in block.ops for value in op.defines}
    blocks = []
    for block in body.blocks:
        ops = []
        for op in block.ops:
            low = ops[-1] if ops else None
            if low is not None and _word(low) and _word(op):
                ref = low.stores[0]
                if replace(ref, addr=ref.addr.plus(2)) == op.stores[0]:
                    whole = mir.extracted_whole(op.args[0], low.args[0], definitions)
                    if whole is not None:
                        ref = replace(ref, width=4)
                        uses = tuple(dict.fromkeys(value for value in (whole.value, ref.base, ref.segment)
                                                  if value is not None))
                        spans = tuple(dict.fromkeys((*low.extra_covers, *op.extra_covers,
                                                     *((op.covers,) if op.covers is not None else ()))))
                        ops[-1] = replace(low, op=ir.Operation.MOVE, name="mov", args=(whole,),
                                          results=(mir.Cell(ref),), stores=(ref,), uses=uses,
                                          merges={}, source_backed=False, raised=None, extra_covers=spans)
                        continue
            ops.append(op)
        blocks.append(replace(block, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))

File: src/optimize/wholestores.rs
//! Store a whole value instead of separately storing its extracted words.
//!
//! Direct port of `qbopt/optimize/wholestores.py`.

use std::collections::BTreeMap;

use crate::model::ir::Operation;
use crate::model::mir::{
    self, Arg, Cell, Kind, MemRef, MirBlock, MirBody, Op, OpCode, OrderedMap, Value,
};

pub(crate) fn _word(op: &Op) -> bool {
    op.kind == Kind::Store
        && !op.barrier()
        && op.defines.is_empty()
        && op.loads.is_empty()
        && op.args.len() == 1
        && op.stores.len() == 1
        && matches!(&op.args[0], Arg::Held(held) if held.width == 2)
        && op.stores[0].width == 2
        && op.stores[0].addr.is_some()
}

pub(crate) fn joined(body: &MirBody) -> MirBody {
    let definitions = body
        .blocks
        .iter()
        .flat_map(|block| &block.ops)
        .flat_map(|op| op.defines.iter().map(move |value| (*value, op)))
        .collect::<BTreeMap<Value, &Op>>();
    let mut blocks = Vec::new();
    for block in &body.blocks {
        let mut ops: Vec<Op> = Vec::new();
        for op in &block.ops {
            if let Some(low) = ops.last() {
                if _word(low) && _word(op) {
                    let r#ref = &low.stores[0];
                    let moved = MemRef {
                        addr: r#ref.addr.map(|addr| addr.plus(2)),
                        ..r#ref.clone()
                    };
                    if moved == op.stores[0] {
                        if let Some(whole) =
                            mir::extracted_whole(&op.args[0], &low.args[0], &definitions)
                        {
                            let r#ref = MemRef {
                                width: 4,
                                ..r#ref.clone()
                            };
                            let mut uses = Vec::new();
                            for value in [Some(whole.value), r#ref.base, r#ref.segment]
                                .into_iter()
                                .flatten()
                            {
                                if !uses.contains(&value) {
                                    uses.push(value);
                                }
                            }
                            let mut absorbed = Vec::new();
                            for one in low.absorbed.iter().chain(&op.absorbed) {
                                if !absorbed.contains(one) {
                                    absorbed.push(*one);
                                }
                            }
                            let replaced = Op {
                                op: Some(OpCode::Operation(Operation::Move)),
                                name: "mov".to_owned(),
                                args: vec![Arg::Held(whole)],
                                results: vec![Arg::Cell(Cell {
                                    r#ref: r#ref.clone(),
                                })],
                                stores: vec![r#ref],
                                uses,
                                merges: OrderedMap::new(),
                                source_backed: false,
                                raised: None,
                                absorbed,
                                ..low.clone()
                            };
                            *ops.last_mut().expect("low") = replaced;
                            continue;
                        }
                    }
                }
            }
            ops.push(op.clone());
        }
        blocks.push(block.with_ops(ops));
    }
    body.with_blocks(blocks)
}
